Automatic logout after changing password on another session

xaben opened the pull request:
| Q | A |
| --- | --- |
| Branch? | 3.1 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17023 |
| License | MIT |
| Doc PR | none |

Although this PR does not break BC, the change in AbstractToken that relates to setUser functionality is highly volatile and should be thoughtfully inspected for possible regressions.

The flow that causes this bug:
1. [ContextListener L161](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Http/Firewall/ContextListener.php#L161) sets the refreshed User in the token (while checking and marking the token as not authenticated [AbstractToken L112](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php#L112)). This as a consequence leads to the fact that the user data from the unserialized session data is lost and cannot be used anymore later on.
2. After Firewall listener trigger, [DaoAuthenticationProvider](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php) is invoked. It fetches the user from the token (the already refreshed user) and runs the checks [UserAuthenticationProvider 85-87](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php#L85).
3. The _checkPreAuth_ [L85](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php#L85) and _checkPostAuth_ [L87](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php#L87) are ok as they use the most recent version of the user (refreshed user). _checkAuthentication_ [L86](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php#L86) on the other hand is intended for comparing the old user (from the token) and the new user (the refreshed one), but as the user was overwritten on step 1, this check always returns true [(DaoAuthenticationProvider checkAuthentication method)](https://github.com/symfony/symfony/blob/3.1/src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php#L56).

As a result the user remains logged in although the serialized data from the session he has, contains an old password. After first refresh of the page, his session data is overwritten with fresh data from the refreshed user.

xaben committed:
- fix user automatic logout after changing password on another session

iltar commented:
You probably have to rebase, you have a bunch of commits in here that are not related to your PR

xaben commented:
Yes sorry rebased it now onto fresh master, the patch was based on current 3.1 branch.

iltar commented on src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php:
```diff
@@ -56,7 +57,7 @@ protected function checkAuthentication(UserInterface $user, UsernamePasswordToke
         if ($currentUser instanceof UserInterface) {
             if ($currentUser->getPassword() !== $user->getPassword()) {
-                throw new BadCredentialsException('The credentials were changed from another session.');
+                throw new AccountPasswordChangedException('The credentials were changed from another session.');
```
As the new Exception extends a different base Exception (account status instead of bad credentials), this is a pretty big BC break

xaben replied:
The exception should extend this other type in order to avoid a redirect loop and unset the wrong(old) token from the session.

Both exceptions are correctly handled, checkPreAuth and checkPostAuth in DaoAuthenticationProvider both throw exceptions that extend this base exception class.

iltar replied:
The `DaoAuthenticationProvider` can be extended or used elsewhere, thus can break user-land implementations :cry:

xaben replied:
What if we remove the token on the BadCredentialsException as well? This change will be done on [ExceptionListener.php L197](https://github.com/symfony/symfony/blob/1ab5be70ae0128c9603401086e5a4727548ccc95/src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php#L197) ??

xaben commented:
@iltar I've changed the PR and removed the custom exception. Instead of it, I added code that removes the token for the case when there is a BadCredentialsException in order to prevent a infinite loop.

fabpot commented on src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php:
```diff
@@ -203,6 +204,15 @@ private function startAuthentication(Request $request, AuthenticationException $
+            $this->tokenStorage->setToken(null);
+
+            if (null !== $this->logger) {
+                $this->logger->info('The security token was removed due to an BadCredentialsException.', array('exception' => $authException));
```
a BadCrendentialsException.

xaben replied:
@fabpot fixed

xabbuh commented on src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php:
```diff
@@ -110,9 +110,9 @@ public function setUser($user)
+            $this->user = $user;
         }
-
-        $this->user = $user;
```
Won't this break applications that rely on the current behaviour?

xaben replied:
Well as I stated at the top of my initial comment, this is the point that is actually the most problematic. And I need input frim somebody that knows really well the security component and it's usage. 

This setter as it stands does too much on it's own, and should be refactored. Most likely separating the logic for marking a token as not authenticated and doing a check elsewhere before setting the token.

The logout fix will not work if we overwrite the user in the token with the refreshed user.

ptrm04 commented:
Should / could this be optional and driven by a security.yml config that defaults to false?

vlechemin commented:
Any update on this? It's an important issue as changing a password to log out any other session is a very common use case and should be the default behaviour on any website.
I think this PR got the two key points right. The only small issue is that the user will be loaded from DB twice but I guess it's acceptable given it's not the general use case.
The other solution would be not to trust an unauthenticated token from the AuthenticationProvider? What would be the purpose of this flag?
Thank you

eXtreme commented:
So after some debugging of mentioned "incorrect" behavior (not being logged out on other sessions after changing the password) I came to the same conclusions: `isAuthenticated` is being correctly set to false after  `hasUserChanged` notices changed password (it starts from ContextListener). But then AccessListener comes into play and reauthenticates the token.
This is clearly not the way it should work and not the way it is mentioned in the docs. 

However, I have not yet tested the patch here.

gharlan commented:
Any update here?

iltar commented:
I don't think this PR should be accepted, imo the flow is working as intended. The current behavior is to log the user out whenever _anything_ changes. This is correct because the user is not the same anymore. If you want to divert from this behavior, implement the [EquatableInterface](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/User/EquatableInterface.php) and customize behavior.

eXtreme commented:
@iltar have you seen my comment above? It **should** log out whenever anything (or, like, a password) changes but it does not.

gharlan commented:
> imo the flow is working as intended. The current behavior is to log the user out whenever anything changes.

No, that is not the current behavior. But I would expect this, too.

See the comment of @eXtreme.

iltar commented:
@eXtreme I've been digging in the security component quite often last weeks and it works exactly as I just described, with the difference that I have the `EquatableInterface` which returns false only when the ID changes. This behavior works for me. Perhaps you can reproduce it in a fork?

eXtreme commented:
@iltar ok I will try later today or tomorrow. The app runs on 2.8 so I will use the same version.

ptrm04 commented:
@iltar have you followed the entire trail of this back to the original issue report? As others mentioned, this is not the current behaviour. https://github.com/symfony/symfony/issues/17023 - might be it also works for ID (can you share your fork), same as for enabled field, but doesn't for other properties.

Security is such an important aspect and shouldn't be ignored - imo this should be picked up by the core team as a high priority (it has been almost 2 years already)

gharlan commented:
@iltar you can use this branch: https://github.com/gharlan/symfony-demo/tree/issue-19033

Go to backend, login, click on the "CHANGE PASSWORD" link.

iltar commented:
I don't have a proper environment to run or test stuff at the moment. Digging into the code, this part looks like the problem to me: https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php#L70-L86
I also found that the refreshUser is never called upon, which I think is normal to use in the other authenticators.

This part in the authenticator, is what causes the problem imo. There's no distinction between the current and actual user, because it always fetches the current user from the current token:
https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php#L77-L80

What _should_ happen imo, is a `refreshUser()` call at that position and return that one. As this behavior is working fine for the Guard implementations for example, I recommend the `AbstractToken` stays the way it is. If anything needs fixing, it's the Dao/User authenticators.

eXtreme commented:
@iltar 
my attempt is here:
https://github.com/eXtreme/symfony-demo/tree/issue-19033
https://github.com/eXtreme/symfony-demo/commit/9f22a02a3f05e0376db4bff1e25a876119c9a662
using a form and a controller etc :P but on 2.8 as I mentioned above, but I guess it's the same on latest branches

simple reproduce:
1. log in as the same user on different browsers
2. in one browser, go to http://localhost:8000/en/profile/change-password
3. change a password to a different one
4. refresh the page on the second browser

I have not managed to reproduce the same behavior I debugged in my real app. BUT. Changing the passwords sets authenticated to false on the second browser. Which... logically makes sense, well, you are not authenticated but still logged in (?). So checking IS_AUTHENTICATED_FULLY should fail right? You can guard your critical app parts by this check and prevent abuse, right? Nope. After adding 

```php
if (!$this->get('security.authorization_checker')->isGranted('IS_AUTHENTICATED_FULLY')) {
    throw $this->createAccessDeniedException();
}
```

to `Admin\BlogController::indexAction` and refreshing the page on the second browser.. you are being reauthenticated. The same thing happens using, for example `@Security("has_role('ROLE_ADMIN') and is_granted('IS_AUTHENTICATED_FULLY')")`. So with these checks for "IS_AUTHENTICATED_FULLY" and changing the password on the first browser and then refreshing the page on the second browser you are reauthenticated. I've put a breakpoint on `AbstractToken::setAuthenticated` and on the second browser there is a call to set it to false and then a call to set it true. The first one is called by `ContextListener::refreshUser`, the second one by `UserAuthenticationProvider`.

In my real app I mentioned in the comment above it was AccessListener which reset the token to authenticated. But maybe it is because I use JMS SecurityExtraBundle? Not sure.

iltar commented:
There's a flag that might alter behavior of the the access decision manager, could that change this particular behavior?

eXtreme commented:
@iltar I'm out of office so I will test my fork tomorrow.

eXtreme commented:
@iltar `consensus` or `unanimous`, it does not change anything in this case.

iltar commented:
@eXtreme I mean this line of code, by default `alwaysAuthenticate` is false, but I have a feeling this is set to true for you: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php#L60

eXtreme commented:
@iltar it's symfony-demo, so it should be default false

gharlan commented:
@iltar I checked this. It is `false`, at least in my demo branch.

iltar commented:
So the issue is really just within the DaoAuthenticator then. What if you would remove just those lines, would that fix the issue as well?

https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php#L77-L80

I see no reason to not always fetch the new data tbh.

eXtreme commented:
@iltar still no luck.

btw. we can move to Slack, not to spam here

iltar commented:
So after some fiddling around with @eXtreme on slack, we have found out the following:
*  `AuthenticationManagerInterface` have to manually implement logic to verify if the token can be authenticated. Only Guard has done this right, because it splits the token in pre and post.

* When a token is deauthenticated because the user has changed, it will trigger a call in the `AuthorizationChecker::isGranted()` method. This will happily reauthenticate the user in the same request as it was just determined that the user had been changed. 

```php
        if ($this->alwaysAuthenticate || !$token->isAuthenticated()) {
            $this->tokenStorage->setToken($token = $this->authenticationManager->authenticate($token));
        }
```

* The `UserAuthenticationProvider` uses a `UsernamePasswordToken`, which is unable to detect whether or not the user has been deauthenticated when going outside of the scope of the `setUser` method.

#### Possible solution
In theory a check could be added to the `ContextListener` which triggers an authentication exception. There's a new exception for this in 2.8 but not in 2.7 (where this problem also exists). There's 1 flaw in the following patch, it doesn't show the users _why_ they have been logged out.

```php
diff --git a/src/Symfony/Component/Security/Http/Firewall/ContextListener.php b/src/Symfony/Component/Security/Http/Firewall/ContextListener.php
index b443fa1..42b950d 100644
--- a/src/Symfony/Component/Security/Http/Firewall/ContextListener.php
+++ b/src/Symfony/Component/Security/Http/Firewall/ContextListener.php
@@ -157,6 +157,12 @@ class ContextListener implements ListenerInterface
                 $refreshedUser = $provider->refreshUser($user);
                 $token->setUser($refreshedUser);

+                // tokens can be deauthenticated if the user has been changed.
+                if (!$token->isAuthenticated()) {
+                    $this->logger->debug('Token was deauthenticated after trying to refresh it.', array('username' => $refreshedUser->getUsername(), 'provider' => get_class($provider)));
+                    return null;
+                }
+
                 if (null !== $this->logger) {
                     $this->logger->debug('User was reloaded from a user provider.', array('username' => $refreshedUser->getUsername(), 'provider' => get_class($provider)));
                 }
```

[This is how Guard solved this](https://github.com/symfony/symfony/blob/6db73d3f4e47fad0be4973b83fb3402bd2321b35/src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php#L69). The `return null` should probably become an exception or something in this patch. @weaverryan perhaps you can confirm this?

weaverryan commented:
@iltar haha, yes, someone finally noticed this in Guard :). I noticed this problem when making Guard, and so I made sure to make Guard not suffer from this. But, it's a workaround in Guard for sure.

About your patch, initially, I like it, but I'd need to review this again in more detail - I can't remember all the moving pieces. In the Guard provider, I throw an instance of `AccountStatusException` because this triggers a "logout" in the `ExceptionListener`: https://github.com/symfony/symfony/blob/3041b4cac1cdce050ea9c7e85b30b8f19fb1bba3/src/Symfony/Component/Security/Http/Firewall/ExceptionListener.php#L199. I *had* to do this, because in that method, I don't have access to the token storage... so I couldn't nullify the token.

But in `ContextListener`, we should just be able to nullify the token in `handle()`. Perhaps:

```php
        if ($token instanceof TokenInterface) {
            $token = $this->refreshUser($token);
        } elseif (null !== $token) {
            if (null !== $this->logger) {
                $this->logger->warning('Expected a security token from the session, got something else.', array('key' => $this->sessionKey, 'received' => $token));
            }
            $token = null;
        }

       // -> right HERE, check if the token is authenticated, and if not, $token = null.
       
        $this->tokenStorage->setToken($token);
```

Hopefully that helps... a bit :)

timwsuqld commented:
Hopefully someone on this PR can answer this clearly. From my investigations, refreshUser isn't being called, and so ANY field changing in a user entity, causes the session to be forced to log back in. But everywhere I'm reading, I'm seeing conflicting information about it should refresh the user, vs logout the user. 
Is this PR fixing it so it correctly reloads the user when a change occurs, or is it also logging it out? I'm trying to use Guards, @weaverryan do I understand correctly that Guards force a logout?

eXtreme commented:
So? Is anyone interested in picking this one up?

weaverryan commented:
I would love if someone could open a PR for this. I don't think this PR is the correct way - it would only fix it for some parts of the system. I think the solution proposed by @iltar and I is a good approach to try: (https://github.com/symfony/symfony/pull/19033#issuecomment-316893991 and comment above).

Anyone want to open a PR?

iltar commented:
@weaverryan I can try, but I'm a bit concerned about the change as it does 2 things:
 - It changes behavior that might break the majority of apps (I can't predict this)
 - It fixes a long standing bug

First point indicates 3.4, second point indicates 2.7, which one should I open it against?

fabpot commented:
I would definitely do this change in 3.4.

chalasr commented:
Closing in favor of https://github.com/symfony/symfony/pull/23882.
